README.md
# react-github-login-button

## Getting Started

`react-github-login-button` is universal, so it can be used client-side or server-side.

1. Install through: `npm install --save react-github-login-button`

2. Import `GithubButton` from `react-github-login-button`:

    ```javascript
    import GithubButton from 'react-github-login-button'
    ```

3. Use `GithubButton` component:

    ```javascript
    <GithubButton
      onClick={() => { console.log('Github button clicked') }}
    />
    ```

## Props

### type
##### PropType
```js
oneOf([ 'light', 'dark' ])
```

##### Default
```js
'dark'
```

##### Example

```js
<GithubButton
  type="light" // can be light or dark
  onClick={() => { console.log('Github button clicked') }}
/>
```

##### Description
`'light'` or `'dark'` for the different github styles (defaults to `dark`)


### disabled
`disabled` - whether or not button is disabled

##### PropType
```js
Boolean
```

##### Default
```js
false
```

##### Example

```javascript
<GithubButton
  disabled // can also be written as disabled={true} for clarity
  onClick={() => { console.log('this will not run on click since it is disabled') }}
/>
```

### label
##### PropType
```js
String
```
##### Default
```js
'Sign in with Github'
```

##### Example

```javascript
<GithubButton
  label='Be Cool'
  onClick={() => { console.log('Github button clicked') }}
/>
```

##### Description
Override the 'Sign in with Github' words with another string.

## Builds

Most commonly people consume `react-github-login-button` as a [CommonJS module](http://webpack.github.io/docs/commonjs.html). This module is what you get when you import redux in a Webpack, Browserify, or a Node environment.

If you don't use a module bundler, it's also fine. The `react-github-login-button` npm package includes precompiled production and development [UMD builds](https://github.com/umdjs/umd) in the [dist folder](https://unpkg.com/react-github-login-button@latest/dist/). They can be used directly without a bundler and are thus compatible with many popular JavaScript module loaders and environments. For example, you can drop a UMD build as a `<script>` tag on the page. The UMD builds make Redux Firestore available as a `window.ReduxFirestore` global variable.

It can be imported like so:

```html
<script src="../node_modules/react-github-login-button/dist/react-github-login-button.min.js"></script>
<!-- or through cdn: <script src="https://unpkg.com/react-github-login-button@latest/dist/react-github-login-button.min.js"></script> -->
<script>console.log('redux firestore:', window.ReactGithubButton)</script>
```

Note: In an effort to keep things simple, the wording from this explanation was modeled after [the installation section of the Redux Docs](https://redux.js.org/#installation).


CHANGELOG.md
# Change Log

This project adheres to [Semantic Versioning](http://semver.org/).
Every release, along with the migration instructions, is documented on the Github [Releases](https://github.com/ehacke/react-github-button/releases) page.


test/unit/index.spec.js
import React from 'react'
import GithubButton from '../../src'
import { shallow } from 'enzyme'

const defaults = {
  label: 'Sign in with Github',
  backgroundColor: '#000000'
}

const disabledStyle = {
  backgroundColor: 'rgba(37, 5, 5, .08)'
}

describe('react-github-button Library', () => {
  describe('exports a component', () => {
    expect(GithubButton).to.exist
  })

  describe('Component', () => {
    let _component

    describe('Default Props', () => {
      beforeEach(() => {
        _component = shallow(<GithubButton />)
      })
      it('Renders div', () => {
        const firstDiv = _component.find('div')
        expect(firstDiv).to.exist
      })

      it('Sets correct label text', () => {
        const firstDiv = _component.find('span')
        expect(firstDiv.first().text()).to.equal(defaults.label)
      })

      it('Applies styles', () => {
        const firstDiv = _component.find('div')
        expect(firstDiv.first()).to.have.style(
          'background-color',
          defaults.backgroundColor
        )
      })
    })

    describe('Custom Label', () => {
      const customLabel = 'Custom Label'

      beforeEach(() => {
        _component = shallow(<GithubButton label={customLabel} />)
      })

      it('Renders div', () => {
        const firstDiv = _component.find('div')
        expect(firstDiv).to.exist
      })

      it('Sets correct label text', () => {
        const firstDiv = _component.find('span')
        expect(firstDiv.first().text()).to.equal(customLabel)
      })
    })

    describe('Disabled', () => {
      beforeEach(() => {
        _component = shallow(<GithubButton disabled />)
      })

      it('Renders div', () => {
        const firstDiv = _component.find('div')
        expect(firstDiv).to.exist
      })

      it('Applies background color', () => {
        const firstDiv = _component.find('div')
        expect(firstDiv.first()).to.have.style(
          'background-color',
          disabledStyle.backgroundColor
        )
      })
    })

    describe('onClick', () => {
      let clickSpy
      beforeEach(() => {
        clickSpy = sinon.spy()
        _component = shallow(<GithubButton onClick={clickSpy} />)
      })

      it('Calls onClick prop when clicked', () => {
        _component.simulate('click')
        expect(clickSpy).to.have.been.calledOnce
      })
    })

    describe('Hover', () => {
      beforeEach(() => {
        _component = shallow(<GithubButton />)
      })

      it('Sets hover style on hover', () => {
        _component.simulate('mouseover')
        const firstDiv = _component.find('div')
        expect(firstDiv.first()).to.have.style(
          'box-shadow',
          '0 0 3px 3px rgba(66,133,244,.3)'
        )
        expect(firstDiv.first()).to.have.style(
          'transition',
          'background-color .218s, border-color .218s, box-shadow .218s'
        )
      })

      it('Sets original style when mouse leave', () => {
        _component.simulate('mouseout')
        const firstDiv = _component.find('div')
        expect(firstDiv.first()).to.have.style(
          'background-color',
          defaults.backgroundColor
        )
        expect(firstDiv.first()).to.have.style(
          'background-color',
          defaults.backgroundColor
        )
      })
    })
  })
})


src/icons.js
import React from 'react'
import PropTypes from 'prop-types'
import {
  iconStyle,
  disabledIconStyle,
  svgStyle,
  disabledSvgStyle
} from './styles'

const svg = (
  <svg
    version="1.2"
    baseProfile="tiny"
    id="Layer_1"
    xmlns="http://www.w3.org/2000/svg"
    x="0px"
    y="0px"
    viewBox="0 0 2350 2314.8"
    xmlSpace="preserve"
    style={svgStyle}
  >
    <path
      d="M1175,0C525.8,0,0,525.8,0,1175c0,552.2,378.9,1010.5,890.1,1139.7c-5.9-14.7-8.8-35.3-8.8-55.8v-199.8H734.4
	c-79.3,0-152.8-35.2-185.1-99.9c-38.2-70.5-44.1-179.2-141-246.8c-29.4-23.5-5.9-47,26.4-44.1c61.7,17.6,111.6,58.8,158.6,120.4
	c47,61.7,67.6,76.4,155.7,76.4c41.1,0,105.7-2.9,164.5-11.8c32.3-82.3,88.1-155.7,155.7-190.9c-393.6-47-581.6-240.9-581.6-505.3
	c0-114.6,49.9-223.3,132.2-317.3c-26.4-91.1-61.7-279.1,11.8-352.5c176.3,0,282,114.6,308.4,143.9c88.1-29.4,185.1-47,284.9-47
	c102.8,0,196.8,17.6,284.9,47c26.4-29.4,132.2-143.9,308.4-143.9c70.5,70.5,38.2,261.4,8.8,352.5c82.3,91.1,129.3,202.7,129.3,317.3
	c0,264.4-185.1,458.3-575.7,499.4c108.7,55.8,185.1,214.4,185.1,331.9V2256c0,8.8-2.9,17.6-2.9,26.4
	C2021,2123.8,2350,1689.1,2350,1175C2350,525.8,1824.2,0,1175,0L1175,0z"
    />
  </svg>
)

const disabledSvg = (
  <svg
    version="1.2"
    baseProfile="tiny"
    id="Layer_1"
    xmlns="http://www.w3.org/2000/svg"
    x="0px"
    y="0px"
    viewBox="0 0 2350 2314.8"
    xmlSpace="preserve"
    style={disabledSvgStyle}
  >
    <path
      d="M1175,0C525.8,0,0,525.8,0,1175c0,552.2,378.9,1010.5,890.1,1139.7c-5.9-14.7-8.8-35.3-8.8-55.8v-199.8H734.4
	c-79.3,0-152.8-35.2-185.1-99.9c-38.2-70.5-44.1-179.2-141-246.8c-29.4-23.5-5.9-47,26.4-44.1c61.7,17.6,111.6,58.8,158.6,120.4
	c47,61.7,67.6,76.4,155.7,76.4c41.1,0,105.7-2.9,164.5-11.8c32.3-82.3,88.1-155.7,155.7-190.9c-393.6-47-581.6-240.9-581.6-505.3
	c0-114.6,49.9-223.3,132.2-317.3c-26.4-91.1-61.7-279.1,11.8-352.5c176.3,0,282,114.6,308.4,143.9c88.1-29.4,185.1-47,284.9-47
	c102.8,0,196.8,17.6,284.9,47c26.4-29.4,132.2-143.9,308.4-143.9c70.5,70.5,38.2,261.4,8.8,352.5c82.3,91.1,129.3,202.7,129.3,317.3
	c0,264.4-185.1,458.3-575.7,499.4c108.7,55.8,185.1,214.4,185.1,331.9V2256c0,8.8-2.9,17.6-2.9,26.4
	C2021,2123.8,2350,1689.1,2350,1175C2350,525.8,1824.2,0,1175,0L1175,0z"
    />
  </svg>
)

export const GithubIcon = ({ disabled, type }) => (
  <div style={!disabled ? iconStyle : { ...iconStyle, ...disabledIconStyle }}>
    {!disabled ? svg : disabledSvg}
  </div>
)

GithubIcon.propTypes = {
  disabled: PropTypes.bool,
  type: PropTypes.oneOf(['light', 'dark'])
}

GithubIcon.defaultProps = {
  type: 'dark'
}


webpack.config.js
'use strict'
const webpack = require('webpack')
const TerserPlugin = require('terser-webpack-plugin')
const pkg = require('./package.json')

const config = {
  module: {
    rules: [
      { test: /\.js$/, loaders: ['babel-loader'], exclude: [/node_modules/] }
    ]
  },
  entry: './src/index.js',
  plugins: [
    new webpack.BannerPlugin({
      banner: 'react-github-button.js v' + pkg.version + ' | (c) ehacke',
      raw: false,
      entryOnly: true
    })
  ],
  externals: {
    react: {
      commonjs: 'react',
      commonjs2: 'react',
      amd: 'react',
      root: 'React'
    },
    'prop-types': {
      commonjs: 'prop-types',
      commonjs2: 'prop-types',
      amd: 'prop-types',
      root: 'PropTypes'
    }
  },
  output: {
    filename: `react-github-button${
      process.env.NODE_ENV === 'production' ? '.min' : ''
    }.js`,
    library: 'ReactGithubButton',
    libraryTarget: 'umd',
    publicPath: '/dist/'
  },
  resolve: {
    extensions: ['.js']
  }
}

if (process.env.NODE_ENV === 'production') {
  config.optimization = {
    minimizer: [
      new TerserPlugin({
        cache: true,
        parallel: true,
        terserOptions: {
          compress: false,
          ecma: 6,
          mangle: true
        },
        sourceMap: true
      })
    ]
  }
}

module.exports = config


src/GithubButton.js
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { GithubIcon } from './icons'
import { darkStyle, lightStyle, disabledStyle, hoverStyle } from './styles'

export default class GithubButton extends PureComponent {
  static propTypes = {
    label: PropTypes.string,
    disabled: PropTypes.bool,
    tabIndex: PropTypes.number,
    onClick: PropTypes.func,
    type: PropTypes.oneOf(['light', 'dark']),
    style: PropTypes.object
  }

  static defaultProps = {
    label: 'Sign in with Github',
    disabled: false,
    type: 'dark',
    tabIndex: 0,
    onClick: () => {}
  }

  state = {
    hovered: false
  }

  getStyle = propStyles => {
    const baseStyle = this.props.type === 'dark' ? darkStyle : lightStyle
    if (this.state.hovered) {
      return { ...baseStyle, ...hoverStyle, ...propStyles }
    }
    if (this.props.disabled) {
      return { ...baseStyle, ...disabledStyle, ...propStyles }
    }
    return { ...baseStyle, ...propStyles }
  }

  mouseOver = () => {
    if (!this.props.disabled) {
      this.setState({ hovered: true })
    }
  }

  mouseOut = () => {
    if (!this.props.disabled) {
      this.setState({ hovered: false })
    }
  }

  click = e => {
    if (!this.props.disabled) {
      this.props.onClick(e)
    }
  }

  render() {
    const { label, style, ...otherProps } = this.props

    return (
      <div
        {...otherProps}
        role="button"
        onClick={this.click}
        style={this.getStyle(style)}
        onMouseOver={this.mouseOver}
        onMouseOut={this.mouseOut}
      >
        <GithubIcon {...this.props} />
        <span>{label}</span>
      </div>
    )
  }
}


src/styles.js
const baseStyle = {
  height: '50px',
  width: '240px',
  border: 'none',
  textAlign: 'center',
  verticalAlign: 'center',
  boxShadow: '0 2px 4px 0 rgba(0,0,0,.25)',
  fontSize: '16px',
  lineHeight: '48px',
  display: 'block',
  borderRadius: '1px',
  transition: 'background-color .218s, border-color .218s, box-shadow .218s',
  fontFamily: 'Roboto,arial,sans-serif',
  cursor: 'pointer',
  userSelect: 'none'
}

export const darkStyle = {
  backgroundColor: '#000000',
  color: '#fff',
  ...baseStyle
}

export const lightStyle = {
  backgroundColor: '#fff',
  color: 'rgba(0,0,0,.54)',
  ...baseStyle
}

export const iconStyle = {
  width: '48px',
  height: '48px',
  justifyContent: 'center',
  alignItems: 'center',
  display: 'flex',
  marginTop: '1px',
  marginLeft: '1px',
  float: 'left',
  backgroundColor: '#fff',
  borderRadius: '1px',
  whiteSpace: 'nowrap'
}

export const svgStyle = {
  width: '24px',
  height: '24px',
  display: 'block'
}

export const disabledSvgStyle = {
  width: '24px',
  height: '24px',
  display: 'block',
  fill: 'rgba(0, 0, 0, .40)'
}

export const hoverStyle = {
  boxShadow: '0 0 3px 3px rgba(66,133,244,.3)',
  transition: 'background-color .218s, border-color .218s, box-shadow .218s'
}

// export const pressedStyle = {
//   backgroundColor: '#3367D6'
// }

export const disabledStyle = {
  backgroundColor: 'rgba(37, 5, 5, .08)',
  color: 'rgba(0, 0, 0, .40)',
  cursor: 'not-allowed'
}

export const disabledIconStyle = {
  backgroundColor: 'transparent'
}
